File: bindings/go/funnelcake_test.go
// Copyright (c) 2020-2026 Kevin Day
//
// SPDX-License-Identifier: BSD-2-Clause-Patent
// See LICENSE.md in the project root for full license text.

package funnelcake

import (
	"errors"
	"testing"
)

// fillByte sets every element of b to v.
func fillByte(b []byte, v byte) {
	for i := range b {
		b[i] = v
	}
}

// fill16 sets every element of b to v.
func fill16(b []uint16, v uint16) {
	for i := range b {
		b[i] = v
	}
}

// TestSDRRoundTrip builds an aligned 256x256 frame, runs a 2x downscale, and
// checks the output geometry. With SIMD available the step must NOT fall back
// to scalar — which proves the frame's planes were correctly aligned.
func TestSDRRoundTrip(t *testing.T) {
	const w, h = 256, 256

	f, err := NewFrame(w, h)
	if err != nil {
		t.Fatalf("NewFrame: %v", err)
	}
	defer f.Close()
	fillByte(f.Y(), 128)
	fillByte(f.U(), 128)
	fillByte(f.V(), 128)

	s, err := NewScaler(ScalerConfig{SrcWidth: w, SrcHeight: h, Flags: Scale2X})
	if err != nil {
		t.Fatalf("NewScaler: %v", err)
	}
	defer s.Close()

	s.Run(f)

	out, ok := s.Output(Scale2X)
	if !ok {
		t.Fatal("2x output was not produced")
	}
	if out.Width != w/2 || out.Height != h/2 {
		t.Errorf("2x output dims = %dx%d, want %dx%d", out.Width, out.Height, w/2, h/2)
	}
	if out.YStride%32 != 0 || out.UVStride%32 != 0 {
		t.Errorf("output strides not 32-aligned: y=%d uv=%d", out.YStride, out.UVStride)
	}
	if len(out.Y()) != out.YStride*out.Height {
		t.Errorf("Y view len = %d, want %d", len(out.Y()), out.YStride*out.Height)
	}
	if SIMDAvailable() {
		if out.Fallback {
			t.Error("SIMD is available but the 2x step fell back to scalar (alignment broken?)")
		}
		if !s.Warnings.Perfect() {
			t.Errorf("expected a perfect result, got warnings scalar=%v partial=%v cropped=%v",
				s.Warnings.Scalar(), s.Warnings.Partial(), s.Warnings.Cropped())
		}
	} else {
		t.Log("SIMD not available on this build/CPU; scalar path exercised")
	}
}

// TestHDRRoundTrip runs a 10-bit I010 source through a 2x step producing both a
// 10-bit HDR output and a tone-mapped SDR output, plus a 1:1 tone-mapped copy.
func TestHDRRoundTrip(t *testing.T) {
	const w, h = 256, 256

	f, err := NewHDRFrame(w, h, PixI010)
	if err != nil {
		t.Fatalf("NewHDRFrame: %v", err)
	}
	defer f.Close()
	fill16(f.Y(), 512)
	fill16(f.U(), 512)
	fill16(f.V(), 512)

	s, err := NewHDRScaler(HDRConfig{
		SrcWidth:  w,
		SrcHeight: h,
		Format:    PixI010,
		Transfer:  TRCPQ,
		Flags:     Scale2X,
		HDRFlags:  Scale2X,
		SDRFlags:  Scale2X,
		Tonemap1x: true,
		Tonemap:   TonemapConfig{Curve: TonemapBT2390},
	})
	if err != nil {
		t.Fatalf("NewHDRScaler: %v", err)
	}
	defer s.Close()

	s.Run(f)

	hdr, ok := s.HDROutput(Scale2X)
	if !ok {
		t.Fatal("2x HDR output not produced")
	}
	if hdr.Width != w/2 || hdr.Height != h/2 {
		t.Errorf("HDR 2x dims = %dx%d, want %dx%d", hdr.Width, hdr.Height, w/2, h/2)
	}
	if len(hdr.Y()) != (hdr.YStride/2)*hdr.Height {
		t.Errorf("HDR Y view len = %d, want %d", len(hdr.Y()), (hdr.YStride/2)*hdr.Height)
	}

	sdr, ok := s.SDROutput(Scale2X)
	if !ok {
		t.Fatal("2x SDR output not produced")
	}
	if sdr.Width != w/2 || sdr.Height != h/2 {
		t.Errorf("SDR 2x dims = %dx%d, want %dx%d", sdr.Width, sdr.Height, w/2, h/2)
	}

	one, ok := s.Tonemap1xOutput()
	if !ok {
		t.Fatal("tonemap_1x output not produced")
	}
	if one.Width != w || one.Height != h {
		t.Errorf("tonemap_1x dims = %dx%d, want %dx%d", one.Width, one.Height, w, h)
	}
}

// assertRowConst checks the first and last output rows are a constant value
// across the active width — catching plane swaps, stride errors, bad pointers.
func assertRowConst(t *testing.T, name string, plane []byte, stride, width, height int, want byte) {
	t.Helper()
	for _, row := range []int{0, height - 1} {
		base := row * stride
		for col := 0; col < width; col++ {
			if plane[base+col] != want {
				t.Fatalf("%s plane row %d col %d = %d, want %d", name, row, col, plane[base+col], want)
				return
			}
		}
	}
}

// TestFlatFieldValues feeds a constant frame and checks each output plane comes
// back as its own constant — exercising the data path, not just dimensions.
func TestFlatFieldValues(t *testing.T) {
	const w, h = 256, 256
	f, err := NewFrame(w, h)
	if err != nil {
		t.Fatal(err)
	}
	defer f.Close()
	fillByte(f.Y(), 128)
	fillByte(f.U(), 64)
	fillByte(f.V(), 192)

	s, err := NewScaler(ScalerConfig{SrcWidth: w, SrcHeight: h, Flags: Scale2X})
	if err != nil {
		t.Fatal(err)
	}
	defer s.Close()
	s.Run(f)

	out, ok := s.Output(Scale2X)
	if !ok {
		t.Fatal("no 2x output")
	}
	assertRowConst(t, "Y", out.Y(), out.YStride, out.Width, out.Height, 128)
	assertRowConst(t, "U", out.U(), out.UVStride, out.Width/2, out.Height/2, 64)
	assertRowConst(t, "V", out.V(), out.UVStride, out.Width/2, out.Height/2, 192)
}

// TestReuseScaler runs the same scaler on two frames.
func TestReuseScaler(t *testing.T) {
	const w, h = 256, 256
	f, _ := NewFrame(w, h)
	defer f.Close()
	s, _ := NewScaler(ScalerConfig{SrcWidth: w, SrcHeight: h, Flags: Scale2X})
	defer s.Close()

	fillByte(f.Y(), 10)
	s.Run(f)
	if out, _ := s.Output(Scale2X); out.Y()[0] != 10 {
		t.Fatalf("first run Y=%d, want 10", out.Y()[0])
	}
	fillByte(f.Y(), 200)
	s.Run(f)
	if out, _ := s.Output(Scale2X); out.Y()[0] != 200 {
		t.Fatalf("second run Y=%d, want 200", out.Y()[0])
	}
}

// TestHDRP010 exercises the semi-planar input path.
func TestHDRP010(t *testing.T) {
	const w, h = 256, 256
	f, err := NewHDRFrame(w, h, PixP010)
	if err != nil {
		t.Fatal(err)
	}
	defer f.Close()
	if f.V() != nil {
		t.Error("P010 should have no separate V plane")
	}

	s, err := NewHDRScaler(HDRConfig{
		SrcWidth: w, SrcHeight: h, Format: PixP010, Transfer: TRCPQ,
		Flags: Scale2X, HDRFlags: Scale2X,
	})
	if err != nil {
		t.Fatal(err)
	}
	defer s.Close()
	s.Run(f)
	if hdr, ok := s.HDROutput(Scale2X); !ok || hdr.Width != w/2 {
		t.Fatalf("P010 2x HDR output wrong: ok=%v", ok)
	}
}

// TestCustomLUTLength rejects a wrong-sized LUT before it reaches C.
func TestCustomLUTLength(t *testing.T) {
	_, err := NewHDRScaler(HDRConfig{
		SrcWidth: 256, SrcHeight: 256, Format: PixI010, Transfer: TRCPQ,
		Flags: Scale2X, SDRFlags: Scale2X,
		Tonemap: TonemapConfig{Curve: TonemapCustom, CustomLUT: make([]byte, 100)},
	})
	if !errors.Is(err, ErrCustomLUTLength) {
		t.Fatalf("err = %v, want ErrCustomLUTLength", err)
	}
}

// TestFrameMismatchPanics confirms a wrong-sized frame is rejected, not read OOB.
func TestFrameMismatchPanics(t *testing.T) {
	s, _ := NewScaler(ScalerConfig{SrcWidth: 256, SrcHeight: 256, Flags: Scale2X})
	defer s.Close()
	f, _ := NewFrame(128, 128)
	defer f.Close()
	defer func() {
		if recover() == nil {
			t.Error("expected a panic on frame/scaler size mismatch")
		}
	}()
	s.Run(f)
}

// TestUseAfterClose confirms a closed scaler panics cleanly instead of passing
// a nil context into C and segfaulting the process.
func TestUseAfterClose(t *testing.T) {
	s, _ := NewScaler(ScalerConfig{SrcWidth: 256, SrcHeight: 256, Flags: Scale2X})
	f, _ := NewFrame(256, 256)
	defer f.Close()
	s.Close()
	defer func() {
		if recover() == nil {
			t.Error("expected a panic when running a closed Scaler")
		}
	}()
	s.Run(f)
}

// TestRunClosedFrame confirms a closed frame is rejected, not passed as null.
func TestRunClosedFrame(t *testing.T) {
	s, _ := NewScaler(ScalerConfig{SrcWidth: 256, SrcHeight: 256, Flags: Scale2X})
	defer s.Close()
	f, _ := NewFrame(256, 256)
	f.Close()
	defer func() {
		if recover() == nil {
			t.Error("expected a panic when running with a closed Frame")
		}
	}()
	s.Run(f)
}

// TestInvalidFlags confirms a hard error surfaces as a typed Error.
func TestInvalidFlags(t *testing.T) {
	// Mixing the thirds (3x) and pow2 (2x) families is invalid.
	_, err := NewScaler(ScalerConfig{SrcWidth: 256, SrcHeight: 256, Flags: Scale2X | Scale3X})
	if err == nil {
		t.Fatal("expected an error for mixed-family flags, got nil")
	}
	if !errors.Is(err, ErrInvalidFlags) {
		t.Errorf("error = %v, want ErrInvalidFlags", err)
	}
}

File: bindings/java/src/org/your/funnelcake/FunnelcakeTest.java
/*
 * Copyright (c) 2020-2026 Kevin Day
 *
 * SPDX-License-Identifier: BSD-2-Clause-Patent
 * See LICENSE.md in the project root for full license text.
 */

package org.your.funnelcake;

import java.lang.foreign.MemorySegment;
import java.util.Optional;

import static java.lang.foreign.ValueLayout.JAVA_BYTE;
import static java.lang.foreign.ValueLayout.JAVA_SHORT;

/**
 * Self-contained smoke test (no JUnit dependency). Run via {@code make test-java}.
 * Exits 0 if all checks pass, 1 otherwise.
 */
public final class FunnelcakeTest {

    private static int failures;

    public static void main(String[] args) {
        sdrRoundTrip();
        hdrRoundTrip();
        invalidFlags();
        flatFieldValues();
        reuseScaler();
        hdrP010();
        customLutLength();
        frameMismatch();

        if (failures == 0) {
            System.out.println("OK  org.your.funnelcake  (all checks passed)");
            System.exit(0);
        } else {
            System.out.println("FAIL  " + failures + " check(s) failed");
            System.exit(1);
        }
    }

    /** 256x256 -> 2x downscale; with SIMD the step must not fall back. */
    private static void sdrRoundTrip() {
        final int w = 256, h = 256;
        try (Frame f = new Frame(w, h);
             Scaler s = new Scaler(new Scaler.Config(w, h, Funnelcake.SCALE_2X))) {
            f.y().fill((byte) 128);
            f.u().fill((byte) 128);
            f.v().fill((byte) 128);

            s.run(f);

            Optional<Output> opt = s.output(Funnelcake.SCALE_2X);
            check("SDR 2x produced", opt.isPresent());
            if (opt.isEmpty()) {
                return;
            }
            Output o = opt.get();
            check("SDR 2x width", o.width() == w / 2);
            check("SDR 2x height", o.height() == h / 2);
            check("SDR y stride 32-aligned", o.yStride() % 32 == 0);
            check("SDR uv stride 32-aligned", o.uvStride() % 32 == 0);
            check("SDR Y plane sized", o.y().byteSize() == (long) o.yStride() * o.height());
            if (Funnelcake.simdAvailable()) {
                check("SDR 2x used SIMD (alignment held)", !o.fallback());
                check("SDR result perfect", s.warnings().perfect());
            } else {
                System.out.println("note: SIMD unavailable; scalar path exercised");
            }
        } catch (Throwable t) {
            fail("SDR round-trip threw: " + t);
        }
    }

    /** 10-bit I010 -> 2x with HDR + tone-mapped SDR + 1:1 copy. */
    private static void hdrRoundTrip() {
        final int w = 256, h = 256;
        HdrScaler.Config cfg = new HdrScaler.Config(
                w, h, Funnelcake.PIX_I010, Funnelcake.TRC_PQ,
                Funnelcake.SCALE_2X, Funnelcake.SCALE_2X, Funnelcake.SCALE_2X,
                0, true, HdrScaler.Tonemap.of(Funnelcake.TONEMAP_BT2390),
                0, false, 0, false);
        try (HdrFrame f = new HdrFrame(w, h, Funnelcake.PIX_I010);
             HdrScaler s = new HdrScaler(cfg)) {

            // Fill luma with a mid 10-bit value.
            long lumaSamples = (long) (f.yStride() / 2) * h;
            MemorySegment y = f.y();
            for (long i = 0; i < lumaSamples; i++) {
                y.setAtIndex(JAVA_SHORT, i, (short) 512);
            }

            s.run(f);

            Optional<HdrOutput> hdr = s.hdrOutput(Funnelcake.SCALE_2X);
            check("HDR 2x produced", hdr.isPresent());
            hdr.ifPresent(o -> {
                check("HDR 2x width", o.width() == w / 2);
                check("HDR 2x height", o.height() == h / 2);
                check("HDR Y plane sized", o.y().byteSize() == (long) o.yStride() * o.height());
            });

            Optional<Output> sdr = s.sdrOutput(Funnelcake.SCALE_2X);
            check("SDR(tone-mapped) 2x produced", sdr.isPresent());
            sdr.ifPresent(o -> {
                check("SDR 2x width", o.width() == w / 2);
                check("SDR 2x height", o.height() == h / 2);
            });

            Optional<Output> one = s.tonemap1xOutput();
            check("tonemap 1x produced", one.isPresent());
            one.ifPresent(o -> {
                check("tonemap 1x width", o.width() == w);
                check("tonemap 1x height", o.height() == h);
            });
        } catch (Throwable t) {
            fail("HDR round-trip threw: " + t);
        }
    }

    /** Mixing thirds and pow2 families is a hard error. */
    private static void invalidFlags() {
        try {
            new Scaler(new Scaler.Config(256, 256, Funnelcake.SCALE_2X | Funnelcake.SCALE_3X)).close();
            fail("expected FunnelcakeException for mixed-family flags");
        } catch (FunnelcakeException e) {
            check("invalid flags -> code -1", e.code() == -1);
        } catch (Throwable t) {
            fail("invalid-flags test threw unexpected: " + t);
        }
    }

    /** Constant input must yield constant output, per-plane — exercises the data path. */
    private static void flatFieldValues() {
        final int w = 256, h = 256;
        try (Frame f = new Frame(w, h);
             Scaler s = new Scaler(new Scaler.Config(w, h, Funnelcake.SCALE_2X))) {
            f.y().fill((byte) 128);
            f.u().fill((byte) 64);
            f.v().fill((byte) 192);
            s.run(f);
            Output o = s.output(Funnelcake.SCALE_2X).orElseThrow();
            checkRowConst("Y", o.y(), o.yStride(), o.width(), o.height(), (byte) 128);
            checkRowConst("U", o.u(), o.uvStride(), o.width() / 2, o.height() / 2, (byte) 64);
            checkRowConst("V", o.v(), o.uvStride(), o.width() / 2, o.height() / 2, (byte) 192);
        } catch (Throwable t) {
            fail("flat-field threw: " + t);
        }
    }

    private static void checkRowConst(String name, MemorySegment p, int stride, int width, int height, byte want) {
        for (int row : new int[]{0, height - 1}) {
            long base = (long) row * stride;
            for (int col = 0; col < width; col++) {
                if (p.get(JAVA_BYTE, base + col) != want) {
                    fail(name + " plane row " + row + " col " + col);
                    return;
                }
            }
        }
        check(name + " plane constant", true);
    }

    /** A scaler can be reused across frames. */
    private static void reuseScaler() {
        final int w = 256, h = 256;
        try (Frame f = new Frame(w, h);
             Scaler s = new Scaler(new Scaler.Config(w, h, Funnelcake.SCALE_2X))) {
            f.y().fill((byte) 10);
            s.run(f);
            check("reuse run #1", s.output(Funnelcake.SCALE_2X).orElseThrow().y().get(JAVA_BYTE, 0) == 10);
            f.y().fill((byte) 200);
            s.run(f);
            check("reuse run #2", (s.output(Funnelcake.SCALE_2X).orElseThrow().y().get(JAVA_BYTE, 0) & 0xFF) == 200);
        } catch (Throwable t) {
            fail("reuse threw: " + t);
        }
    }

    /** Semi-planar P010 path produces correctly-sized output. */
    private static void hdrP010() {
        final int w = 256, h = 256;
        try (HdrFrame f = new HdrFrame(w, h, Funnelcake.PIX_P010);
             HdrScaler s = new HdrScaler(new HdrScaler.Config(
                     w, h, Funnelcake.PIX_P010, Funnelcake.TRC_PQ,
                     Funnelcake.SCALE_2X, Funnelcake.SCALE_2X, 0))) {
            s.run(f);
            HdrOutput o = s.hdrOutput(Funnelcake.SCALE_2X).orElseThrow();
            check("P010 2x dims", o.width() == w / 2 && o.height() == h / 2);
        } catch (Throwable t) {
            fail("P010 threw: " + t);
        }
    }

    /** A wrong-sized custom LUT is rejected before it reaches C. */
    private static void customLutLength() {
        try {
            new HdrScaler(new HdrScaler.Config(
                    256, 256, Funnelcake.PIX_I010, Funnelcake.TRC_PQ,
                    Funnelcake.SCALE_2X, 0, Funnelcake.SCALE_2X,
                    0, false,
                    new HdrScaler.Tonemap(Funnelcake.TONEMAP_CUSTOM, 0, 0,
                            Funnelcake.RANGE_LIMITED, Funnelcake.RANGE_LIMITED, new byte[100]),
                    0, false, 0, false)).close();
            fail("expected IllegalArgumentException for short custom LUT");
        } catch (IllegalArgumentException e) {
            check("custom LUT length rejected", true);
        } catch (Throwable t) {
            fail("custom LUT test threw unexpected: " + t);
        }
    }

    /** A mismatched frame size is rejected, not read out of bounds. */
    private static void frameMismatch() {
        try (Scaler s = new Scaler(new Scaler.Config(256, 256, Funnelcake.SCALE_2X));
             Frame f = new Frame(128, 128)) {
            s.run(f);
            fail("expected IllegalArgumentException for frame/scaler mismatch");
        } catch (IllegalArgumentException e) {
            check("frame mismatch rejected", true);
        } catch (Throwable t) {
            fail("frame mismatch test threw unexpected: " + t);
        }
    }

    private static void check(String name, boolean ok) {
        if (ok) {
            System.out.println("  pass: " + name);
        } else {
            fail(name);
        }
    }

    private static void fail(String name) {
        failures++;
        System.out.println("  FAIL: " + name);
    }
}
